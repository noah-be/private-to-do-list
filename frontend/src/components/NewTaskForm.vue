<template>
  <div id="new-task-form">
    <input
      type="text"
      id="task-input"
      v-model="task"
      placeholder="Enter task..."
    />
    <select id="priority-input" v-model="priority">
      <option value="Low">Low</option>
      <option value="Medium">Medium</option>
      <option value="High">High</option>
    </select>
    <button id="add-task-btn" @click="addTask">Add Task</button>
  </div>
</template>

<script>
export default {
  data() {
    return {
      task: "",
      priority: "Medium",
    };
  },
  methods: {
    addTask() {
      // TODO: Handle adding a new task
    },
  },
};
</script>
