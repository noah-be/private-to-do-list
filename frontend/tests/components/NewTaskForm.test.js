import { describe, it, expect, vi } from "vitest";
import { mount } from "@vue/test-utils";
import NewTaskForm from "@/components/NewTaskForm.vue";

describe("NewTaskForm.vue", () => {
  it("matches the snapshot", () => {
    const wrapper = mount(NewTaskForm);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("binds input fields correctly", async () => {
    const wrapper = mount(NewTaskForm);
    const titleInput = wrapper.find("#task-input");
    const dueDateInput = wrapper.find("#due-date-input");
    const prioritySelect = wrapper.find("#priority-input");

    await titleInput.setValue("Test Task");
    await dueDateInput.setValue("2024-12-01");
    await prioritySelect.setValue("High");

    expect(wrapper.vm.title).toBe("Test Task");
    expect(wrapper.vm.dueDate).toBe("2024-12-01");
    expect(wrapper.vm.priority).toBe("High");
  });

  it("calls fetch with correct data", async () => {
    global.fetch = vi.fn(() =>
      Promise.resolve({
        ok: true,
        json: () =>
          Promise.resolve({
            title: "Test Task",
            priority: "High",
            dueDate: "2024-12-01",
          }),
      }),
    );

    const wrapper = mount(NewTaskForm);

    await wrapper.find("#task-input").setValue("Test Task");
    await wrapper.find("#priority-input").setValue("High");
    await wrapper.find("#due-date-input").setValue("2024-12-01");
    await wrapper.find("#add-task-btn").trigger("click");

    expect(global.fetch).toHaveBeenCalledWith(
      "/api/task",
      expect.objectContaining({
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${localStorage.getItem("token")}`,
        },
        body: JSON.stringify({
          title: "Test Task",
          priority: "High",
          dueDate: "2024-12-01",
        }),
      }),
    );
  });

  it("emits task-added with the new task", async () => {
    const mockTask = {
      title: "Test Task",
      priority: "High",
      dueDate: "2024-12-01",
    };

    global.fetch = vi.fn(() =>
      Promise.resolve({
        ok: true,
        json: () => Promise.resolve(mockTask),
      }),
    );

    const wrapper = mount(NewTaskForm);

    await wrapper.find("#task-input").setValue("Test Task");
    await wrapper.find("#priority-input").setValue("High");
    await wrapper.find("#due-date-input").setValue("2024-12-01");
    await wrapper.find("#add-task-btn").trigger("click");

    expect(wrapper.emitted("task-added")).toBeTruthy();
    expect(wrapper.emitted("task-added")[0][0]).toEqual(mockTask);
  });

  it("shows an alert if task title or due date is missing", async () => {
    const alertMock = vi.spyOn(window, "alert").mockImplementation(() => {});
    const wrapper = mount(NewTaskForm);

    await wrapper.find("#add-task-btn").trigger("click");

    expect(alertMock).toHaveBeenCalledWith(
      "Task title and due date are required.",
    );
    alertMock.mockRestore();
  });

  it("resets the form after adding a task", async () => {
    const wrapper = mount(NewTaskForm);

    global.fetch = vi.fn(() =>
      Promise.resolve({
        ok: true,
        json: () => Promise.resolve({}),
      }),
    );

    await wrapper.find("#task-input").setValue("Test Task");
    await wrapper.find("#due-date-input").setValue("2024-12-01");
    await wrapper.find("#add-task-btn").trigger("click");

    expect(wrapper.vm.title).toBe("");
    expect(wrapper.vm.priority).toBe("Medium");
    expect(wrapper.vm.dueDate).toBe("");
  });
});
