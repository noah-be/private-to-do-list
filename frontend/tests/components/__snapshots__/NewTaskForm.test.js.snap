// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NewTaskForm.vue > matches the snapshot 1`] = `
"<div id="new-task-form"><input type="text" id="task-input" placeholder="Enter task..."><select id="priority-input">
    <option value="Low">Low</option>
    <option value="Medium">Medium</option>
    <option value="High">High</option>
  </select><input type="date" id="due-date-input"><button id="add-task-btn">Add Task</button></div>"
`;
